#!/usr/bin/env nextflow
nextflow.enable.dsl=2

process bbmerge{
    //conda "/gpfs/projects/bgmp/shared/groups/2025/chimera/envs/getBarcodes" --new bbtools version
    conda "/gpfs/projects/bgmp/shared/groups/2025/chimera/envs/old_bbmerge" // --old bbtools version
    input:
        tuple val(sampleId), path(read1), path(read2)

    output:
        path "${sampleId}_merged.fastq"

    publishDir "/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/", mode: 'copy'

    script:
    """
    echo $read1
    echo $read2
    /usr/bin/time -v /gpfs/projects/bgmp/shared/groups/2025/chimera/envs/old_bbmerge/bbtools/lib/bbmerge.sh \
    in1=${read1}\
    in2=${read2}\
    out=${sampleId}_merged.fastq \
    outu1=/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/${sampleId}1_unmerged.fastq \
    outu2=/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/${sampleId}2_unmerged.fastq\
    ihist=/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/${sampleId}_bbmerge_stats.txt
    """
}

process hts_primers{
    conda "/gpfs/projects/bgmp/shared/groups/2025/chimera/envs/getBarcodes"
    input:
        path merged_file
    output:
        path "${merged_file.simpleName}_flip_trim_SE.fastq.gz"
    
    publishDir "/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/", mode: 'copy'

    script:
    """
    /usr/bin/time -v hts_Primers -U ${merged_file} -f ${merged_file.simpleName}_flip_trim -P /projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/IlluminaTestData/pSG1stub_FWD.fasta -Q /projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/IlluminaTestData/pSG1stub_REV.fasta -l 5 -x -e 6 -d 6 -L /projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/G01_test_flip_stats.txt -F
    """
}

process make_barcode_fasta {
    input:
        path trimmed_file
    output:
        path "${trimmed_file.simpleName}.cut.txt"

    publishDir "/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/", mode: 'copy'

    script:
    // need to "escape" special characters like "$". Otherwise, nextflow thinks this is an object in this workflow, not within the script.

    //remember to change the "20" to "24" when running our data !!!!!
    """
    /usr/bin/time -v gzip -cd ${trimmed_file} \
    | awk 'NR%4 ==2' | cut -c1-20 | awk 'BEGIN{FIELDWIDTHS="20"} {print \$1}' > ${trimmed_file.simpleName}.cut.txt 
    """
}

process count_barcodes{
    input:
        path barcode_file
    output:
        path "${barcode_file}.count.txt"
    publishDir "/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/", mode: 'copy'

    script:
    //fix the if the first barcode is empty check here to just be 1 command
    //added grep command to remove empty lines before unique counting
    //this should remove "empty" barcode counts that are generated by unique
    //this could be parallelized further later with find and xargs, but it means making a file or a temp file
    """
    /usr/bin/time -v cat ${barcode_file} | sort -T ./ --parallel=40 | grep -v "^\$" | uniq -c | \
    awk '{print \$2 "\t" \$1}' > ${barcode_file}.count.txt
    """
}

process star_code{
    conda "/projects/bgmp/shared/groups/2025/chimera/envs/starcode_env"
    input:
        path count_file

    output:
        path "${count_file.simpleName}_collapse_d1.tsv"
    
    publishDir "/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/test_outfiles/", mode: 'copy'
    
    script:
    """
    /usr/bin/time -v starcode -d1 --sphere -i ${count_file} --output "${count_file.simpleName}_collapse_d1.tsv" --print-clusters
    """
}

workflow {
    // Create channels for paired reads
    read_pairs = Channel
        .fromFilePairs('/projects/bgmp/shared/groups/2025/chimera/mlscha/bgmp-2025-chimera/IlluminaTestData/*_R{1,2}.fastq')
        .map { sampleId, reads -> tuple(sampleId, reads[0], reads[1]) }

    //run bbmerge (note: need to make a nextflow.config file that has "conda.enabled = true" to activate existing envs in this nextflow file --done!)
    bbmerge(read_pairs)

    //collect the "output" file paths (only the merged file)
    merged_files = bbmerge.out
    merged_files.view()

    //run HTS Primers
    hts_primers(merged_files)

    trimmed_file = hts_primers.out
    //make the fasta file with all barcodes
    make_barcode_fasta(trimmed_file)

    //grab barcode file
    barcode_file = make_barcode_fasta.out

    //count barcodes from barcode file
    count_barcodes(barcode_file)

    //grab counted barcode file
    count_file = count_barcodes.out
    star_code(count_file)


}

