#!/bin/bash
#SBATCH --account=bgmp
#SBATCH --partition=bgmp
#SBATCH --mem=64GB
#SBATCH --cpus-per-task=8
#SBATCH --job-name=pacbio
#SBATCH --out=/projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/results/slurm/slurm-%j.out
#SBATCH --err=/projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/results/slurm/slurm-%j.err

set -e

# Input BAM file to process
input_file="/projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/input_files/test.bam"
# Input reference genes file (FASTA) to use for alignment
input_ref_genes_file="/projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/input_files/dcus_mut.genes"
# Input reference proteins file (FAA) to use for postprocessing
input_ref_proteins_file="/projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/input_files/dcus_mut.proteins"
# Output directory to publish results
output_dir="/projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/results/"

run_iter_num=$(ls -1 -d $output_dir/*/ | sed -r 's/.+_(.+)\//\1/' | sort -n | awk '{num=$1} END{print num+1}')
results_dir="${output_dir}/results_${run_iter_num}"
mkdir $results_dir
input_file_basename=$(basename $input_file .bam)

envs_dir="/projects/bgmp/shared/groups/2025/chimera/envs/"

tmp_dir_root=$(mktemp -d)


# Step 1: Convert from BAM to FASTQ
step_name="01_bedtools"
tmp_dir_bedtools="${tmp_dir_root}/${step_name}/"
mkdir $tmp_dir_bedtools
bedtools_output="${tmp_dir_bedtools}/${input_file_basename}.fastq"
mamba run -p "${envs_dir}/bedtools" /usr/bin/time -v bamToFastq \
    -i $input_file \
    -fq $bedtools_output
# Compress file, update output name to reflect the change
gzip $bedtools_output
bedtools_output="${tmp_dir_bedtools}/${input_file_basename}.fastq.gz"
cp -r $tmp_dir_bedtools "${results_dir}/${step_name}"


# Step 2: Extract barcodes and sequences
step_name="02_extract_seqs"
tmp_dir_extract_seqs="${tmp_dir_root}/${step_name}/"
mkdir $tmp_dir_extract_seqs
# This file is specified on the command line
extract_seqs_fasta_output="${tmp_dir_extract_seqs}/${input_file_basename}.fasta"
# This file is generated by the script as a side effect
extract_seqs_tsv_output="${tmp_dir_extract_seqs}/${input_file_basename}bc_stats_for_starcode.tsv"
mamba run -p "${envs_dir}/biopython" /usr/bin/time -v python /projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/barcode_extract_v9_no_coll_noEnvZ_HK_MP.NEW.py \
    -i $bedtools_output \
    -o $extract_seqs_fasta_output
# This file output name is also modified as a side effect
extract_seqs_fasta_output="${tmp_dir_extract_seqs}/${input_file_basename}_noN.fasta"
cp -r $tmp_dir_extract_seqs "${results_dir}/${step_name}"


# Step 3: Starcode
step_name="03_starcode"
tmp_dir_starcode="${tmp_dir_root}/${step_name}/"
mkdir $tmp_dir_starcode
starcode_output="${tmp_dir_starcode}/${input_file_basename}.tsv"
mamba run -p "${envs_dir}/starcode_env" /usr/bin/time -v starcode \
    -d1 \
    --sphere \
    --print-clusters \
    -i $extract_seqs_tsv_output \
    --output $starcode_output
cp -r $tmp_dir_starcode "${results_dir}/${step_name}"


# Step 4: Call consensus
step_name="04_call_consensus"
tmp_dir_call_consensus="${tmp_dir_root}/${step_name}/"
mkdir $tmp_dir_call_consensus
call_consensus_output="${tmp_dir_call_consensus}/${input_file_basename}.fasta"
mamba run -p "${envs_dir}/r-consensus-call" /usr/bin/time -v Rscript /projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/collision_dcuS.NEW.R \
    $tmp_dir_call_consensus \
    $extract_seqs_fasta_output \
    $starcode_output \
    $call_consensus_output
cp -r $tmp_dir_call_consensus "${results_dir}/${step_name}"


# Step 5: Alignment
step_name="05_alignment"
tmp_dir_alignment="${tmp_dir_root}/${step_name}/"
mkdir $tmp_dir_alignment
alignment_mapped_output="${tmp_dir_alignment}/${input_file_basename}.map.sam"
alignment_unmapped_output="${tmp_dir_alignment}/${input_file_basename}.unmap.sam"
# Index the reference
reference_file="${tmp_dir_alignment}/reference.fasta"
cp $input_ref_genes_file $reference_file
mamba run -p "${envs_dir}/old_bbmerge" /usr/bin/time -v bbmap.sh \
    ref=$reference_file
# Align to the reference
mamba run -p "${envs_dir}/old_bbmerge" /usr/bin/time -v mapPacBio.sh \
    in=$call_consensus_output \
    outm=$alignment_mapped_output \
    outu=$alignment_unmapped_output
cp -r $tmp_dir_alignment "${results_dir}/${step_name}"


# Step 6: Translate DNA (technically, this could be run in parallel with the alignment)
step_name="06_translate"
tmp_dir_translate="${tmp_dir_root}/${step_name}/"
mkdir $tmp_dir_translate
translate_output="${tmp_dir_translate}/${input_file_basename}.aa.csv"
mamba run -p "${envs_dir}/biopython" /usr/bin/time -v python /projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/process_alt_trans.py \
    $call_consensus_output \
    $translate_output
cp -r $tmp_dir_translate "${results_dir}/${step_name}"


# Step 7: Summarize
step_name="07_summarize"
tmp_dir_summarize="${tmp_dir_root}/${step_name}/"
mkdir $tmp_dir_summarize
mamba run -p "${envs_dir}/biopython" /usr/bin/time -v python /projects/bgmp/shared/groups/2025/chimera/jonat/bgmp-2025-chimera/original_pacbio_pipeline/parse_sam_script.py \
    $input_ref_proteins_file \
    $translate_output \
    $alignment_mapped_output \
    $tmp_dir_summarize
cp -r $tmp_dir_summarize "${results_dir}/${step_name}"

exit
